README
NAME
    Dancer2::Session::Sereal - Dancer 2 session storage in files with Sereal

VERSION
    version 0.001

DESCRIPTION
    This module implements Dancer 2 session engine based on Sereal files.

    This backend can be used in single-machine production environments, but
    two things should be kept in mind: The content of the session files is
    not encrypted or protected in anyway and old session files should be
    purged by a CRON job.

CONFIGURATION
    The setting session should be set to "Sereal" in order to use this
    session engine in a Dancer2 application.

    Files will be stored to the value of the setting "session_dir", whose
    default value is "appdir/sessions".

    Here is an example configuration that use this session engine and stores
    session files in /tmp/dancer-sessions

        session: "Sereal"

        engines:
          session:
            Sereal:
              session_dir: "/tmp/dancer-sessions"

SUPPORT
  Bugs / Feature Requests
    Please report any bugs or feature requests through the issue tracker at
    <https://github.com/dagolden/dancer2-session-sereal/issues>. You will be
    notified automatically of any progress on your issue.

  Source Code
    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    <https://github.com/dagolden/dancer2-session-sereal>

      git clone git://github.com/dagolden/dancer2-session-sereal.git

AUTHOR
    David Golden <dagolden@cpan.org>

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2013 by David Golden.

    This is free software, licensed under:

      The Apache License, Version 2.0, January 2004



Changes
Revision history for Dancer2-Session-Sereal

{{$NEXT}}

0.005     2020-05-26 18:47:34-04:00 America/New_York
    - Issues getting 0.004 to release. Bump to 0.005 (cromedome)
    
0.004 (unreleased)
  [ API CHANGES ]
    - None

  [ BUG FIXES ]
    - None

  [ DOCUMENTATION ]
    - Document 'encoder_args' and 'decoder_args' so tests pass.

  [ ENHANCEMENTS ]
    - None

  [ NEW FEATURES ]
    - None

  [ STATISTICS ]
    - N/A

0.003 2016-11-05
  [ NEW FEATURES ]
    - Add 'encoder_args' and 'decoder_args' configuration options. (GH#3)

  [ STATISTICS ]
    - code churn: 3 files changed, 113 insertions(+), 12 deletions(-)

0.002 2016-08-16
  - Convert tests from Test::TCP to Plack::Test
  - Minimum Dancer2 release is now 0.150000

  [ STATISTICS ]
    - code churn: 3 files changed, 232 insertions(+), 167 deletions(-)

0.001 2013-02-22T23:40:14Z America/New_York
  - First release


lib/Dancer2/Session/Sereal.pm
use 5.008001;
use strict;
use warnings;

package Dancer2::Session::Sereal;
# ABSTRACT: Dancer 2 session storage in files with Sereal
# VERSION

use Moo;
use Dancer2::Core::Types;
use Sereal::Encoder;
use Sereal::Decoder;

our $VERSION = '0.006';

#--------------------------------------------------------------------------#
# Attributes
#--------------------------------------------------------------------------#

has _suffix => (
    is      => 'ro',
    isa     => Str,
    default => sub { ".srl" },
);

=head2 encoder_args

Arguments to be passed to L<Sereal::Encoder>. If these are not provided in
the session configuration, sensible defaults are used.

=cut

has encoder_args => (
    is => 'ro',
    isa => HashRef,
    default => sub{
        +{
            snappy         => 1,
            croak_on_bless => 1,
        }
    }
);

has _encoder => (
    is      => 'lazy',
    isa     => InstanceOf ['Sereal::Encoder'],
    handles => { '_freeze' => 'encode' },
);

sub _build__encoder {
    my ($self) = @_;
    return Sereal::Encoder->new( $self->encoder_args );
}

=head2 decoder_args

Arguments to be passed to L<Sereal::Decoder>. If these are not provided in
the session configuration, sensible defaults are used.

=cut

has decoder_args => (
    is => 'ro',
    isa => HashRef,
    default => sub{
        +{
            refuse_objects => 1,
            validate_utf8  => 1,
        }
    }
);

has _decoder => (
    is      => 'lazy',
    isa     => InstanceOf ['Sereal::Decoder'],
    handles => { '_thaw' => 'decode' },
);

sub _build__decoder {
    my ($self) = @_;
    return Sereal::Decoder->new( $self->decoder_args );
}

#--------------------------------------------------------------------------#
# Role composition
#--------------------------------------------------------------------------#

with 'Dancer2::Core::Role::SessionFactory::File';

sub _freeze_to_handle {
  my ($self, $fh, $data) = @_;
  binmode $fh;
  print {$fh} $self->_freeze($data);
  return;
}

sub _thaw_from_handle {
  my ($self, $fh) = @_;
  binmode($fh);
  return $self->_thaw( do { local $/; <$fh> } );
}

1;
__END__

=head1 DESCRIPTION

This module implements Dancer 2 session engine based on L<Sereal> files.

This backend can be used in single-machine production environments, but two
things should be kept in mind: The content of the session files is not
encrypted or protected in anyway and old session files should be purged by a
CRON job.

=head1 CONFIGURATION

The setting B<session> should be set to C<Sereal> in order to use this session
engine in a Dancer2 application.

Files will be stored to the value of the setting C<session_dir>, whose default
value is C<appdir/sessions>.

Arguments for the L<Sereal::Encoder> and L<Sereal::Decoder> objects can be 
given via the C<encoder_args> and C<decoder_args>. If not provided, they default to
C<< snappy => 1, croak_on_bless =>1 >> and C<< refuse_objects => 1, validate_utf8 => 1 >>, respectively.

Here is an example configuration that use this session engine and stores session
files in /tmp/dancer-sessions

    session: "Sereal"

    engines:
      session:
        Sereal:
          session_dir: "/tmp/dancer-sessions"
          encoder_args:
            snappy:         1
            croak_on_bless: 1
          decoder_args:
            refuse_objects: 1
            validate_utf8:  1


=cut

# vim: ts=4 sts=4 sw=4 et:


Makefile.PL
use strict;
use warnings;

use 5.008001;

use ExtUtils::MakeMaker 6.17;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Dancer 2 session storage in files with Sereal",
  "AUTHOR" => "David Golden <dagolden\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.17"
  },
  "DISTNAME" => "Dancer2-Session-Sereal",
  "LICENSE" => "apache",
  "MIN_PERL_VERSION" => "5.008001",
  "NAME" => "Dancer2::Session::Sereal",
  "PREREQ_PM" => {
    "Dancer2" => "0.15000",
    "Dancer2::Core::Role::SessionFactory::File" => 0,
    "Dancer2::Core::Types" => 0,
    "Moo" => 0,
    "Sereal::Decoder" => 0,
    "Sereal::Encoder" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Dancer2" => "0.15000",
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "File::Temp" => "0.22",
    "HTTP::Cookies" => 0,
    "HTTP::Date" => 0,
    "HTTP::Request::Common" => 0,
    "JSON" => 0,
    "Plack::Test" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.006",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Dancer2" => "0.15000",
  "Dancer2::Core::Role::SessionFactory::File" => 0,
  "Dancer2::Core::Types" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "File::Temp" => "0.22",
  "HTTP::Cookies" => 0,
  "HTTP::Date" => 0,
  "HTTP::Request::Common" => 0,
  "JSON" => 0,
  "Moo" => 0,
  "Plack::Test" => 0,
  "Sereal::Decoder" => 0,
  "Sereal::Encoder" => 0,
  "Test::More" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
